File: withu/Networking/APIClient.swift
//
//  APIClient.swift
//  withu
//

import Foundation

enum APIError: Error, LocalizedError {
    case invalidResponse
    case server(status: Int, detail: String)
    case decoding(Error)
    case transport(Error)
    /// 402 — 무료/크레딧 소진. balance 는 갱신된 잔액(있으면).
    case paymentRequired(balance: Entitlement?)

    var errorDescription: String? {
        switch self {
        case .invalidResponse:
            return "서버 응답 형식이 잘못됐어요."
        case .server(let status, let detail):
            return "서버 오류 \(status): \(detail)"
        case .decoding(let err):
            return "디코딩 실패: \(err.localizedDescription)"
        case .transport(let err):
            return "통신 실패: \(err.localizedDescription)"
        case .paymentRequired:
            return "무료 횟수를 다 썼어요."
        }
    }
}

/// 모든 Error 를 사용자 친화 한국어 메시지로 변환.
/// catch 블록의 `error.localizedDescription` 대신 사용.
extension Error {
    var koreanizedDescription: String {
        if let api = self as? APIError {
            switch api {
            case .invalidResponse:
                return "서버 응답이 이상해요. 잠시 후 다시 시도해 주세요."
            case .server(let status, _):
                if status == 429 { return "요청이 너무 많아요. 잠시 후 다시 시도해 주세요." }
                if status >= 500 { return "서버에 문제가 생겼어요. 잠시 후 다시 시도해 주세요." }
                if status == 422 { return "프롬프트가 안전 정책에 걸렸어요. 단어를 살짝 바꿔서 다시 시도해 주세요." }
                return "서버 오류 (\(status)). 잠시 후 다시 시도해 주세요."
            case .decoding:
                return "결과를 읽을 수 없어요. 다시 시도해 주세요."
            case .transport(let inner):
                return (inner as Error).koreanizedDescription
            case .paymentRequired:
                return "무료 횟수를 다 썼어요. 충전하거나 구독해 주세요."
            }
        }
        if let urlErr = self as? URLError {
            switch urlErr.code {
            case .notConnectedToInternet:
                return "인터넷에 연결돼 있지 않아요. Wi-Fi 또는 셀룰러를 확인해 주세요."
            case .timedOut:
                return "응답이 너무 오래 걸려요. 잠시 후 다시 시도해 주세요."
            case .cannotConnectToHost, .cannotFindHost:
                return "서버에 연결할 수 없어요. 네트워크 또는 서버 상태를 확인해 주세요."
            case .networkConnectionLost:
                return "연결이 끊겼어요. 다시 시도해 주세요."
            case .cancelled:
                return "요청이 취소됐어요."
            default:
                return "네트워크 오류가 발생했어요. 다시 시도해 주세요."
            }
        }
        // 기본 — 시스템 로컬라이즈된 메시지 (영문일 수 있음) 보다 깔끔한 한국어 폴백
        return "오류가 발생했어요. 잠시 후 다시 시도해 주세요."
    }
}

actor APIClient {
    static let shared = APIClient()

    private let session: URLSession
    private let decoder: JSONDecoder
    private let encoder: JSONEncoder

    private init() {
        let config = URLSessionConfiguration.default
        config.timeoutIntervalForRequest = APIConfig.timeout
        config.timeoutIntervalForResource = APIConfig.resourceTimeout
        config.waitsForConnectivity = true   // 잠깐 끊기면 죽이지 말고 기다리기
        self.session = URLSession(configuration: config)

        self.decoder = JSONDecoder()
        self.decoder.keyDecodingStrategy = .convertFromSnakeCase

        self.encoder = JSONEncoder()
        self.encoder.keyEncodingStrategy = .convertToSnakeCase
    }

    func ping() async throws -> Bool {
        let url = APIConfig.baseURL.appendingPathComponent("/health")
        do {
            let (_, response) = try await session.data(from: url)
            guard let http = response as? HTTPURLResponse else {
                throw APIError.invalidResponse
            }
            return (200..<300).contains(http.statusCode)
        } catch let error as APIError {
            throw error
        } catch {
            throw APIError.transport(error)
        }
    }

    /// 빠른 reachability 체크 — 5초 timeout, 별도 ephemeral session.
    /// 기존 `session` 은 30분 timeout + waitsForConnectivity 라서 연결 끊김 시 한참 매달림.
    /// generate 전에 이 메서드로 먼저 확인하면 연결 안 됐을 때 즉시 에러.
    func preflightPing() async throws {
        let url = APIConfig.baseURL.appendingPathComponent("/health")
        let config = URLSessionConfiguration.ephemeral
        config.timeoutIntervalForRequest = 5
        config.timeoutIntervalForResource = 5
        config.waitsForConnectivity = false
        let fastSession = URLSession(configuration: config)
        do {
            let (_, response) = try await fastSession.data(from: url)
            guard let http = response as? HTTPURLResponse,
                  (200..<300).contains(http.statusCode) else {
                throw APIError.invalidResponse
            }
        } catch let error as APIError {
            throw error
        } catch {
            throw APIError.transport(error)
        }
    }

    // MARK: - 인증 (Phase 2)

    /// Sign in with Apple identityToken 으로 서버 세션 발급.
    func authenticateApple(identityToken: String) async throws -> AppleAuthResponse {
        let url = APIConfig.baseURL.appendingPathComponent("/auth/apple")
        var req = URLRequest(url: url)
        req.httpMethod = "POST"
        req.setValue("application/json", forHTTPHeaderField: "Content-Type")
        req.httpBody = try encoder.encode(AppleAuthRequest(identityToken: identityToken))
        return try await send(req, decode: AppleAuthResponse.self)
    }

    /// 계정 삭제 (Bearer 필요). 서버의 이 사용자 데이터 전체 삭제.
    func deleteAccount() async throws {
        let url = APIConfig.baseURL.appendingPathComponent("/me")
        var req = URLRequest(url: url)
        req.httpMethod = "DELETE"
        req.timeoutInterval = 30   // 삭제는 짧게 — 생성용 session 의 30분 timeout 상속 방지
        if let sessionToken = KeychainStore.sessionToken() {
            req.setValue("Bearer \(sessionToken)", forHTTPHeaderField: "Authorization")
        }
        try await sendDiscardingBody(req)
    }

    /// 서버 권리 스냅샷 (Bearer 필요).
    func fetchMe() async throws -> Entitlement {
        let url = APIConfig.baseURL.appendingPathComponent("/me")
        var req = URLRequest(url: url)
        if let sessionToken = KeychainStore.sessionToken() {
            req.setValue("Bearer \(sessionToken)", forHTTPHeaderField: "Authorization")
        }
        return try await send(req, decode: MeResponse.self).entitlement
    }

    /// StoreKit 결제 JWS 를 서버에 제출해 멱등 적립 (Bearer 필요).
    func verifyPurchase(signedTransaction: String) async throws -> Entitlement {
        let url = APIConfig.baseURL.appendingPathComponent("/iap/verify")
        var req = URLRequest(url: url)
        req.httpMethod = "POST"
        req.setValue("application/json", forHTTPHeaderField: "Content-Type")
        if let sessionToken = KeychainStore.sessionToken() {
            req.setValue("Bearer \(sessionToken)", forHTTPHeaderField: "Authorization")
        }
        req.httpBody = try encoder.encode(IapVerifyRequest(signedTransaction: signedTransaction))
        return try await send(req, decode: MeResponse.self).entitlement
    }

    /// 할인코드 적용 (Bearer 필요).
    func redeem(code: String) async throws -> Entitlement {
        try await postCode(path: "/redeem", body: RedeemRequest(code: code))
    }

    /// 친구 추천코드 적용 (Bearer 필요).
    func applyReferral(code: String) async throws -> Entitlement {
        try await postCode(path: "/referral/apply", body: ReferralRequest(code: code))
    }

    private func postCode<B: Encodable>(path: String, body: B) async throws -> Entitlement {
        let url = APIConfig.baseURL.appendingPathComponent(path)
        var req = URLRequest(url: url)
        req.httpMethod = "POST"
        req.setValue("application/json", forHTTPHeaderField: "Content-Type")
        if let sessionToken = KeychainStore.sessionToken() {
            req.setValue("Bearer \(sessionToken)", forHTTPHeaderField: "Authorization")
        }
        req.httpBody = try encoder.encode(body)
        return try await send(req, decode: MeResponse.self).entitlement
    }

    /// 공통 요청 → 본문 무시 (성공 status 만 확인). 계정 삭제 등 빈 응답용.
    private func sendDiscardingBody(_ request: URLRequest) async throws {
        do {
            let (data, response) = try await session.data(for: request)
            guard let http = response as? HTTPURLResponse else { throw APIError.invalidResponse }
            guard (200..<300).contains(http.statusCode) else {
                let detail = (try? decoder.decode(APIErrorDetail.self, from: data))?.detail
                    ?? String(data: data, encoding: .utf8) ?? ""
                throw APIError.server(status: http.statusCode, detail: detail)
            }
        } catch let error as APIError {
            throw error
        } catch {
            throw APIError.transport(error)
        }
    }

    /// 공통 요청 → 디코드.
    private func send<T: Decodable>(_ request: URLRequest, decode: T.Type) async throws -> T {
        do {
            let (data, response) = try await session.data(for: request)
            guard let http = response as? HTTPURLResponse else { throw APIError.invalidResponse }
            guard (200..<300).contains(http.statusCode) else {
                let detail = (try? decoder.decode(APIErrorDetail.self, from: data))?.detail
                    ?? String(data: data, encoding: .utf8) ?? ""
                throw APIError.server(status: http.statusCode, detail: detail)
            }
            do { return try decoder.decode(T.self, from: data) }
            catch { throw APIError.decoding(error) }
        } catch let error as APIError {
            throw error
        } catch {
            throw APIError.transport(error)
        }
    }

    /// 이미지 생성. kind: "single"|"batch" (서버 무료 버킷 구분), batchId: 일괄 세션 묶음.
    func generateImage(_ request: GenerateImageRequest,
                       kind: String = "single",
                       batchId: String? = nil) async throws -> GenerateImageResponse {
        let url = APIConfig.baseURL.appendingPathComponent("/generate")
        var req = URLRequest(url: url)
        req.httpMethod = "POST"
        req.setValue("application/json", forHTTPHeaderField: "Content-Type")
        if let token = APIConfig.apiToken, !token.isEmpty {
            req.setValue(token, forHTTPHeaderField: "X-Withu-Token")
        }
        if let sessionToken = KeychainStore.sessionToken() {
            req.setValue("Bearer \(sessionToken)", forHTTPHeaderField: "Authorization")
        }
        // 재시도 이중차감 방지 키 + 무료 버킷 구분
        req.setValue(UUID().uuidString, forHTTPHeaderField: "Idempotency-Key")
        req.setValue(kind, forHTTPHeaderField: "X-Withu-Kind")
        if let batchId {
            req.setValue(batchId, forHTTPHeaderField: "X-Withu-Batch")
        }
        req.httpBody = try encoder.encode(request)

        do {
            let (data, response) = try await session.data(for: req)
            guard let http = response as? HTTPURLResponse else {
                throw APIError.invalidResponse
            }

            // 402 — 무료/크레딧 소진
            if http.statusCode == 402 {
                let balance = (try? decoder.decode(PaymentRequiredResponse.self, from: data))?.balance
                throw APIError.paymentRequired(balance: balance)
            }

            guard (200..<300).contains(http.statusCode) else {
                let detail = (try? decoder.decode(APIErrorDetail.self, from: data))?.detail
                    ?? String(data: data, encoding: .utf8)
                    ?? ""
                throw APIError.server(status: http.statusCode, detail: detail)
            }

            do {
                return try decoder.decode(GenerateImageResponse.self, from: data)
            } catch {
                throw APIError.decoding(error)
            }
        } catch let error as APIError {
            throw error
        } catch {
            throw APIError.transport(error)
        }
    }
}

File: android/app/src/main/java/com/seoyoung/withu/net/ApiClient.kt
package com.seoyoung.withu.net

import com.seoyoung.withu.BuildConfig
import kotlinx.coroutines.CancellationException
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.delay
import kotlinx.coroutines.withContext
import kotlinx.serialization.json.Json
import okhttp3.MediaType.Companion.toMediaType
import okhttp3.OkHttpClient
import okhttp3.Request
import okhttp3.RequestBody.Companion.toRequestBody
import java.util.UUID
import java.util.concurrent.TimeUnit

/**
 * withu API 클라이언트 — iOS APIClient.swift 포팅 (스펙 11).
 * 같은 Cloudflare Worker 사용. 인증은 X-Withu-Token 경로만 (로그인은 SCOPE 제외 — Bearer 미전송).
 * 이미지 생성이 medium 1–3분/high 2–5분 — read timeout 30분/전체 60분.
 * 짧게 줄이면 진행 중 생성이 잘못된 "offline" 으로 죽는다 (iOS 교훈).
 */
object ApiClient {
    private const val BASE_URL = "https://withu-api.ysy1398.workers.dev"

    /** preflight 재시도 — iOS APIClient.preflight 와 동일 (2회 시도 · 0.7초 간격). */
    private const val PREFLIGHT_ATTEMPTS = 2
    private const val PREFLIGHT_RETRY_DELAY_MS = 700L

    // encodeDefaults: steps/width/height/quality/style/model 기본값도 서버로 전송 (iOS 파리티).
    // explicitNulls=false: null 필드는 본문에서 생략 (iOS JSONEncoder 의 nil 생략과 동일).
    private val json = Json { ignoreUnknownKeys = true; encodeDefaults = true; explicitNulls = false }

    private val client = OkHttpClient.Builder()
        .connectTimeout(15, TimeUnit.SECONDS)
        .readTimeout(30, TimeUnit.MINUTES)     // iOS timeout=1800
        .callTimeout(60, TimeUnit.MINUTES)     // iOS resourceTimeout=3600
        .build()

    /**
     * preflight 전용 8초 클라이언트 — 본 세션은 30분 timeout 이라 연결이 끊겼을 때
     * 한참 매달림. generate 직전 이걸로 즉시 실패 판정 (iOS ephemeral 세션 대응).
     * 5초는 화면 재진입 직후 등 일시적 지연에서 오탐이 나서 8초로 완화 (iOS 파리티).
     */
    private val preflightClient = client.newBuilder()
        .connectTimeout(8, TimeUnit.SECONDS)
        .readTimeout(8, TimeUnit.SECONDS)
        .callTimeout(8, TimeUnit.SECONDS)
        .build()

    private val jsonMediaType = "application/json".toMediaType()

    /**
     * 이미지 생성. 402 → ApiError.PaymentRequired 로 분류.
     * @param kind X-Withu-Kind — "single"|"batch" (서버 무료 버킷 구분)
     * @param batchId X-Withu-Batch — 일괄 세션 묶음 (배치일 때만)
     * @param idempotencyKey 재시도 이중차감 방지 — 배치는 job.id 고정 전달
     */
    suspend fun generateImage(
        req: GenerateImageRequest,
        kind: String = "single",
        batchId: String? = null,
        sessionId: String? = null,
        state: String? = null,
        idempotencyKey: String = UUID.randomUUID().toString(),
    ): GenerateImageResponse = withContext(Dispatchers.IO) {
        val body = json.encodeToString(GenerateImageRequest.serializer(), req)
            .toRequestBody(jsonMediaType)
        val builder = Request.Builder()
            .url("$BASE_URL/generate")
            .post(body)
            .header("Content-Type", "application/json")
            .header("Idempotency-Key", idempotencyKey)
            .header("X-Withu-Kind", kind)
            // 생성 모니터링 메타 (서버 gen_events 로깅용 — 수정 체인/상태/플랫폼)
            .header("X-Withu-Platform", "android")
        val token = BuildConfig.WITHU_API_TOKEN
        if (token.isNotEmpty()) builder.header("X-Withu-Token", token)
        if (batchId != null) builder.header("X-Withu-Batch", batchId)
        if (sessionId != null) builder.header("X-Withu-Session", sessionId)
        if (state != null) builder.header("X-Withu-State", state)

        try {
            client.newCall(builder.build()).execute().use { resp ->
                val text = resp.body?.string() ?: ""
                // 402 — 무료/크레딧 소진
                if (resp.code == 402) {
                    val balance = runCatching {
                        json.decodeFromString<PaymentRequiredResponse>(text)
                    }.getOrNull()?.balance
                    throw ApiError.PaymentRequired(balance)
                }
                if (!resp.isSuccessful) {
                    throw ApiError.Server(resp.code, detailOf(text))
                }
                try {
                    json.decodeFromString<GenerateImageResponse>(text)
                } catch (e: Exception) {
                    throw ApiError.Decoding(e.message ?: "unknown")
                }
            }
        } catch (e: ApiError) {
            throw e   // ApiError 는 그대로 재던지기 — Transport 로 이중 래핑 금지 (스펙 11 §4.1)
        } catch (e: Exception) {
            throw ApiError.Transport(e)
        }
    }

    /**
     * 빠른 reachability 체크 — /health, 8초 × 2회 (실패 후 0.7초 대기). 2xx 아니면 ApiError throw.
     * 1회 시도는 화면 재진입 직후의 일시적 실패에서 "연결이 어려워요" 오탐이 났다 (iOS 파리티).
     */
    suspend fun preflightPing(): Unit = withContext(Dispatchers.IO) {
        var last: Exception = ApiError.InvalidResponse()
        repeat(PREFLIGHT_ATTEMPTS) { attempt ->
            try {
                val req = Request.Builder().url("$BASE_URL/health").build()
                preflightClient.newCall(req).execute().use { resp ->
                    if (!resp.isSuccessful) throw ApiError.InvalidResponse()
                }
                return@withContext
            } catch (e: CancellationException) {
                throw e   // 취소는 재시도 대상이 아니다 — 호출부가 취소를 따로 처리한다
            } catch (e: Exception) {
                last = e
                if (attempt < PREFLIGHT_ATTEMPTS - 1) delay(PREFLIGHT_RETRY_DELAY_MS)
            }
        }
        throw if (last is ApiError) last else ApiError.Transport(last)
    }

    /** /health — 기본 세션으로 Bool 반환 (실패는 false). */
    suspend fun health(): Boolean = withContext(Dispatchers.IO) {
        runCatching {
            val req = Request.Builder().url("$BASE_URL/health").build()
            client.newCall(req).execute().use { it.isSuccessful }
        }.getOrDefault(false)
    }

    /**
     * Google 로그인 핸드셰이크 — POST /auth/google { id_token } → 세션 토큰 + entitlement.
     * (iOS /auth/apple 대응. 서버가 계정 키를 "google:<sub>" 로 네임스페이스.)
     */
    suspend fun authGoogle(idToken: String): AuthResponse = withContext(Dispatchers.IO) {
        val bodyJson = json.encodeToString(AuthGoogleRequest.serializer(), AuthGoogleRequest(idToken))
        try {
            val req = Request.Builder()
                .url("$BASE_URL/auth/google")
                .post(bodyJson.toRequestBody(jsonMediaType))
                .header("Content-Type", "application/json")
                .build()
            client.newCall(req).execute().use { resp ->
                val text = resp.body?.string() ?: ""
                if (!resp.isSuccessful) throw ApiError.Server(resp.code, detailOf(text))
                try {
                    json.decodeFromString<AuthResponse>(text)
                } catch (e: Exception) {
                    throw ApiError.Decoding(e.message ?: "unknown")
                }
            }
        } catch (e: ApiError) {
            throw e
        } catch (e: Exception) {
            throw ApiError.Transport(e)
        }
    }

    /**
     * 할인코드 적용 — POST /redeem.
     * 로그인 제외 범위라 서버가 401 을 줄 수 있음 — 에러는 그대로 throw (caller 가 koreanized).
     */
    suspend fun redeem(code: String): Entitlement =
        postCode("/redeem", json.encodeToString(RedeemRequest.serializer(), RedeemRequest(code)))

    /** 친구 초대 코드 적용 — POST /referral/apply. */
    suspend fun applyReferral(code: String): Entitlement =
        postCode("/referral/apply", json.encodeToString(ReferralRequest.serializer(), ReferralRequest(code)))

    private suspend fun postCode(path: String, bodyJson: String): Entitlement =
        withContext(Dispatchers.IO) {
            try {
                val req = Request.Builder()
                    .url("$BASE_URL$path")
                    .post(bodyJson.toRequestBody(jsonMediaType))
                    .header("Content-Type", "application/json")
                    .build()
                client.newCall(req).execute().use { resp ->
                    val text = resp.body?.string() ?: ""
                    if (!resp.isSuccessful) {
                        throw ApiError.Server(resp.code, detailOf(text))
                    }
                    try {
                        json.decodeFromString<MeResponse>(text).entitlement
                    } catch (e: Exception) {
                        throw ApiError.Decoding(e.message ?: "unknown")
                    }
                }
            } catch (e: ApiError) {
                throw e
            } catch (e: Exception) {
                throw ApiError.Transport(e)
            }
        }

    /** 비 2xx 본문 → detail 필드 우선, 실패 시 raw body (앞부분만). */
    private fun detailOf(text: String): String =
        runCatching { json.decodeFromString<ApiErrorDetail>(text).detail }
            .getOrElse { text.take(300) }
}
